github.com/cockroachdb/pebble/internal/metamorphic: TestMetaTwoInstance failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMetaTwoInstance [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/17578359?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/17578359?buildTab=artifacts#meta) on refs/heads/master @ [71bb6ba29095](https://github.com/cockroachdb/pebble/commits/71bb6ba29095):


```
        iter20.SeekLT("exlxiaw@3", "") // [false] <nil> #3138
        iter20.Last() // [true,"omgc",<no point>,["omgc","tusyvljv@1")=>{""="kndcvqcsuegp","@9"="rnnrfzfycyppxbyywvn"}*] <nil> #3166
        iter20.SeekGE("jilwthysnty@7", "") // [true,"leytqraym@9",<no point>,["leytqraym@9","ndjxwnqxiykn")=>{""="kndcvqcsuegp","@9"="ppobfjw"}*] <nil> #3176
        iter20.SetBounds("", "jdpkkpywvk@7") // <nil> #3182
        // 05:26:12.926 INFO: [JOB 70] compacted(default) L0 [000038 000049] (5.4KB) Score=100.00 + L6 [000042] (1.6KB) Score=0.00 -> L6 [000050] (2.7KB), in 0.0s (0.0s total), output rate 544KB/s
        iter30.Prev("") // [true,"kqbgbhegr",<no point>,["kqbgbhegr","mcptdj@9")=>{""="kndcvqcsuegp","@9"="ppobfjw","@8"="fijop","@4"="egcekdmwyko"}*] <nil> #3257
        iter30.SetBounds("pvtdfmhftakb@1", "udty@7") // <nil> #3259
        iter20.SeekLT("jdpkkpywvk@7", "") // [true,"hxylycxmyft",<no point>,["hxylycxmyft","jdpkkpywvk@7")=>{"@9"="yjhlwayywydotfkv"}*] <nil> #3183
        iter21.Next("") // [false] <nil> #3132
        // 05:26:12.927 INFO: [JOB 108] compacting: sstable created 000071
        iter22.Next("") // [true,"hphdgdyyf",<no point>,["hphdgdyyf","hxylycxmyft")=>{""="ccjtckcyocobv"}*] <nil> #2987
        iter22.Next("") // [true,"hxylycxmyft",<no point>,["hxylycxmyft","jilwthysnty@7")=>{"@9"="yjhlwayywydotfkv"}*] <nil> #3003
        // 05:26:12.929 INFO: [JOB 72] WAL deleted 000047
        db2.LogData("u") // <nil> #3062
        db1.Delete("wqusxjuldlk@6") // <nil> #3016
        iter29.Last() // [true,"yfibrmosts",<no point>,["yfibrmosts","yzbbrcrl")=>{"@7"="bqxwmdfxcqzulrv","@3"="msuxzothagfdqgxtyele"}*] <nil> #3216
        iter21.SetBounds("", "hpnqr@6") // <nil> #3170
        iter20.Next("") // [false] <nil> #3184
        // 05:26:12.929 INFO: [JOB 108] compacted(elision-only) L6 [000070] (2.3KB) Score=0.00 + L6 [] (0B) Score=0.00 -> L6 [000071] (1.5KB), in 0.0s (0.0s total), output rate 302KB/s
        iter20.Prev("") // [true,"hxylycxmyft",<no point>,["hxylycxmyft","jdpkkpywvk@7")=>{"@9"="yjhlwayywydotfkv"}*] <nil> #3185
        iter21.SeekLT("hpnqr@6", "") // [true,"hpbhhsryttfj",<no point>,["hpbhhsryttfj","hpnqr@6")=>{"@9"="yjhlwayywydotfkv"}*] <nil> #3171
        // 05:26:12.930 INFO: [JOB 108] sstable deleted 000070
        iter20.InternalNext() // switching from reverse to forward via internalNext is prohibited #3188
        iter21.Prev("") // [true,"hgdvmaju@5","axedhzixv",["exlxiaw","hpbhhsryttfj")=>{"@9"="yjhlwayywydotfkv","@7"="odgiylhedf"}*] <nil> #3172
        iter20.SeekGE("bzuwaamepf@6", "") // [true,"hphdgdyyf",<no point>,["hphdgdyyf","hxylycxmyft")=>{""="ccjtckcyocobv"}*] <nil> #3193
        iter20.Next("") // [true,"hxylycxmyft",<no point>,["hxylycxmyft","jdpkkpywvk@7")=>{"@9"="yjhlwayywydotfkv"}*] <nil> #3204
        iter20.SeekGE("mcptdj@2", "") // [false] <nil> #3209
        iter21.SeekGE("jkbz@9", "pmhwbicl@7") // [invalid] <nil> #3174
        iter20.First() // [true,"afjmkhwqj",<no point>,["afjmkhwqj","awzouxzoybm")=>{""="ccjtckcyocobv"}*] <nil> #3213
        iter20.SetBounds("viifwwi", "yzbbrcrl@8") // <nil> #3236
        iter21.First() // [true,"aizpxbgfaahp@7","tidahnnicfsirhtp",<no range>] <nil> #3178
        iter20.SeekGE("hpnqr@6", "") // [true,"viifwwi",<no point>,["viifwwi","wmzivwx")=>{""="kndcvqcsuegp","@9"="rnnrfzfycyppxbyywvn"}*] <nil> #3251
        iter20.SetBounds("nzyjipu@3", "ofijhd@9") // <nil> #3262
        iter21.Prev("tusyvljv@3") // [invalid] <nil> #3205
        iter20.SeekLT("ofijhd@9", "") // [true,"nzyjipu@3",<no point>,["nzyjipu@3","ofijhd@9")=>{""="kndcvqcsuegp"}*] <nil> #3263
        iter20.Next("") // [false] <nil> #3264
        iter20.Next("") // [false] <nil> #3265
        iter20.Prev("") // [true,"nzyjipu@3",<no point>,["nzyjipu@3","ofijhd@9")=>{""="kndcvqcsuegp"}*] <nil> #3266
        iter21.First() // [true,"aizpxbgfaahp@7","tidahnnicfsirhtp",<no range>] <nil> #3206
        db1.Delete("hpbhhsryttfj@2") // <nil> #3021
        iter21.Prev("") // [false] <nil> #3208
        iter21.NextPrefix() // [false] NextPrefix not permitted with upper bound hpnqr@6 #3219
        iter29.SetBounds("bwkoyxadgpw@6", "wldfbinnn@8") // <nil> #3228
        db1.Get("pvtdfmhftakb@1") // [""] pebble: not found #3024
        db1.Delete("kqbgbhegr") // <nil> #3033
        iter21.SeekLT("iidpz@9", "") // [true,"hpbhhsryttfj",<no point>,["hpbhhsryttfj","hpnqr@6")=>{"@9"="yjhlwayywydotfkv"}*] <nil> #3225
        iter30.SeekGE("pvtdfmhftakb@1", "") // [true,"pvtdfmhftakb@1",<no point>,["pvtdfmhftakb@1","sqnjppip")=>{"@9"="rnnrfzfycyppxbyywvn","@8"="fijop","@4"="egcekdmwyko"}*] <nil> #3260
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMetaTwoInstance$' --run-dir /artifacts/meta/241104-052554.3303162653962/standard-027 --try-to-reduce -v
        --- FAIL: TestMetaTwoInstance/execution/standard-027 (2.94s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMetaTwoInstance$' ./internal/metamorphic -seed 1730697954330043506 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMetaTwoInstance.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-294

--- a/sstable/reader_iter_single_lvl.go
+++ b/sstable/reader_iter_single_lvl.go
@@ -283,15 +283,6 @@ func newRowBlockSingleLevelIterator(
 	)
 	if r.tableFormat >= TableFormatPebblev3 {
 		if r.Properties.NumValueBlocks > 0 {
-			// NB: we cannot avoid this ~248 byte allocation, since valueBlockReader
-			// can outlive the singleLevelIterator due to be being embedded in a
-			// LazyValue. This consumes ~2% in microbenchmark CPU profiles, but we
-			// should only optimize this if it shows up as significant in end-to-end
-			// CockroachDB benchmarks, since it is tricky to do so. One possibility
-			// is that if many sstable iterators only get positioned at latest
-			// versions of keys, and therefore never expose a LazyValue that is
-			// separated to their callers, they can put this valueBlockReader into a
-			// sync.Pool.
 			i.vbReader = valueBlockReader{
 				bpOpen: i,
 				rp:     rp,

--- a/sstable/value_block.go
+++ b/sstable/value_block.go
@@ -731,13 +731,11 @@ type valueBlockReader struct {
 	rp     ReaderProvider
 	vbih   valueBlocksIndexHandle
 	stats  *base.InternalIteratorStats
+
 	// fetcher is allocated lazily the first time we create a LazyValue, in order
 	// to avoid the allocation if we never read a lazy value (which should be the
 	// case when we're reading the latest value of a key).
 	fetcher *valueBlockFetcher
-
-	// lazyFetcher is the LazyFetcher value embedded in any LazyValue that was returned.
-	lazyFetcher base.LazyFetcher
 }
 
 var _ block.GetLazyValueForPrefixAndValueHandler = (*valueBlockReader)(nil)
@@ -753,9 +751,9 @@ func (r *valueBlockReader) GetLazyValueForPrefixAndValueHandle(handle []byte) ba
 		// contain an array of instances, a subset of which have been given out).
 		r.fetcher = newValueBlockFetcher(r.bpOpen, r.rp, r.vbih, r.stats)
 	}
-	fetcher := &r.lazyFetcher
+	lazyFetcher := &r.fetcher.lazyFetcher
 	valLen, h := decodeLenFromValueHandle(handle[1:])
-	*fetcher = base.LazyFetcher{
+	*lazyFetcher = base.LazyFetcher{
 		Fetcher: r.fetcher,
 		Attribute: base.AttributeAndLen{
 			ValueLen:       int32(valLen),
@@ -768,7 +766,7 @@ func (r *valueBlockReader) GetLazyValueForPrefixAndValueHandle(handle []byte) ba
 	}
 	return base.LazyValue{
 		ValueOrHandle: h,
-		Fetcher:       fetcher,
+		Fetcher:       lazyFetcher,
 	}
 }
 
@@ -803,6 +801,10 @@ type valueBlockFetcher struct {
 	valueCache    block.BufferHandle
 	closed        bool
 	bufToMangle   []byte
+
+	// lazyFetcher is the LazyFetcher value embedded in any LazyValue that we
+	// return. It is used to avoid having a separate allocation for that.
+	lazyFetcher base.LazyFetcher
 }
 
 var _ base.ValueFetcher = (*valueBlockFetcher)(nil)

